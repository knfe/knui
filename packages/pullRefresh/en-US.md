<!-- English -->

# kn-pullrefresh

### Install

```js
import KnPullRefresh from 'kn-ui'

Vue.use(KnPullRefresh)
```

### Usage

#### Basic Usage

The refresh event will be triggered when pull refresh, you should set v-model to false to reset loading status after process refresh event.

```html
<van-pull-refresh v-model="isLoading" @refresh="onRefresh"></van-pull-refresh>
```

```js
export default {
  data() {
    return {
      isLoading: false
    }
  },

  methods: {
    onRefresh() {
      setTimeout(() => {
        this.isLoading = false
      }, 500)
    }
  }
}
```

### API

| Attribute          | Description               | Type      | Default               | 版本 |
| ------------------ | ------------------------- | --------- | --------------------- | ---- |
| v-model            | Loading status            | `Boolean` | -                     | -    |
| pulling-text       | Text to show when pulling | `String`  | `Pull to refresh...`  | -    |
| loosing-text       | Text to show when loosing | `String`  | `Loose to refresh...` | -    |
| loading-text       | Text to show when loading | `String`  | `Loading...`          | -    |
| animation-duration | Animation duration        | `Number`  | `300`                 | -    |
| head-height        | Height of head            | `Number`  | `50`                  | -    |
| disabled           | Whether to disable        | `Boolean` | `false`               | -    |

### Event

| Event   | Description                 | Parameters |
| ------- | --------------------------- | ---------- |
| refresh | Triggered when pull refresh | -          |

### Slot

| name    | Description                           |
| ------- | ------------------------------------- |
| normal  | Content of head when at normal status |
| pulling | Content of head when at pulling       |
| loosing | Content of head when at loosing       |
| loading | Content of head when at loading       |
