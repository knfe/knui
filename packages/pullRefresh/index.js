import KnPullRefresh from './src/main.vue'

KnPullRefresh.install = function(Vue) {
  Vue.component(KnPullRefresh.name, KnPullRefresh)
}

export default KnPullRefresh
