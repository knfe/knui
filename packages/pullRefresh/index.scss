/**
 * kn-pullrefresh component style
 */
@import '../style/index.scss';

.kn-pullrefresh {
  overflow: hidden;
  user-select: none; // 用户不能选中复制文本

  .track {
    position: relative;
  }

  .head {
    position: absolute;
    top: -50px;
    left: 0;
    width: 100%;
    height: 50px;
    overflow: hidden;
    font-size: 14px;
    line-height: 50px;
    color: gray;
    text-align: center;

    .hint {
      width: 100%;
      height: 50px;
      line-height: 50px;
      text-align: center;
    }
  }
}
