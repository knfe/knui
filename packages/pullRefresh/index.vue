
<template>
  <div class="kn-pullrefresh">
    <div
      class="track"
      @touchstart="onTouchStart"
      @touchmove="onTouchMove"
      @touchend="onTouchEnd"
      @touchcancel="onTouchEnd"
      :style="style"
    >
      <!-- 下拉刷新的部分 -->
      <div class="head" :style="setHeadStyle">
        <!-- 正常状态 -->
        <slot v-if="status === 'normal'" name="normal"/>
        <!-- 下拉状态 -->
        <slot v-if="status === 'pulling'" name="pulling">
          <div class="hint" :style="setHintStyle">{{pullingText}}</div>
        </slot>
        <!-- 松手状态 -->
        <slot v-if="status === 'loosing'" name="loosing">
          <div class="hint" :style="setHintStyle">{{loosingText}}</div>
        </slot>
        <!-- 加载状态 -->
        <slot v-if="status === 'loading'" name="loading">
          <div class="hint" :style="setHintStyle">{{loadingText}}</div>
        </slot>
      </div>
      <!-- 页面内容 -->
      <slot/>
    </div>
  </div>
</template>
<script>
/**
 * @function 下拉刷新组件
 * @module packages/pullRefresh
 * @param {Boolean} [disabled]是否禁用功能
 * @param {String} [pullingText = '下拉即可刷新']下拉时状态文本
 * @param {String} [loosingText = '释放即可刷新']释放前状态文本
 * @param {Boolean} [loadingText = '正在加载...']加载中状态文本
 * @param {Number} [headHeight = 50]状态文本头部高度
 * @param {Number} [animationDuration = 300]动画时长
 * @param {Boolean} value - 必传
 * @example
 * <kn-pull-refresh v-model="isLoading" @refresh="onRefresh"></kn-pull-refresh>
 */

// import { getScroll } from '../utils/shim.js'

export default {
  name: 'kn-pull-refresh',
  data() {
    return {
      duration: 0, // 动画时间
      height: 0, // 移动距离
      startY: 0, // 起始位置
      status: 'normal', // 状态 - normal: 未操作常态, pulling: 下拉状态, loosing: 即将松手状态, loading: 加载状态
      invalid: 0, // 是否禁用
      setHeadStyle: '', // 设置头部样式
      setHintStyle: '' // 设置提示文本样式
    }
  },
  mounted() {
    this.setHeadStyle = `height: ${this.headHeight}px; line-height: ${this.headHeight}px; top: -${this.headHeight}px;`
    this.setHintStyle = `height: ${this.headHeight}px; line-height: ${this.headHeight}px;`
  },
  props: {
    disabled: Boolean, // 是否禁用下拉刷新
    pullingText: {
      type: String,
      default: '下拉即可刷新'
    },
    loosingText: {
      type: String,
      default: '释放即可刷新'
    },
    loadingText: {
      type: String,
      default: '正在加载...'
    },
    headHeight: {
      type: Number,
      default: 50
    },
    animationDuration: {
      type: Number,
      default: 300
    },
    value: {
      type: Boolean,
      required: true
    }
  },
  computed: {
    /**
     * 页面样式
     */
    style() {
      return {
        transition: `${this.duration}ms`,
        transform: `translate3d(0,${this.height}px, 0)`
      }
    },
    /**
     * 是否为禁用状态
     */
    untouchable() {
      return this.status === 'loading' || this.disabled || this.invalid
    }
  },
  methods: {
    onTouchStart(e) {
      // 触摸开始时，未在顶部，此次触摸标记为无效
      this.invalid = this.getScrollTop() > 0 ? 1 : 0
      if (this.untouchable) return
      this.duration = 0
      this.startY = e.touches[0].clientY
    },
    onTouchMove(e) {
      if (this.untouchable) return
      if (e.touches[0].clientY - this.startY > 0) {
        this.getStatus(Math.sqrt((e.touches[0].clientY - this.startY) * 40))
        e.preventDefault() // 阻止事件冒泡
      }
    },
    onTouchEnd(e) {
      if (this.untouchable) return
      // 判断如果移动的距离小于headHeight高度，则不刷新
      if (e.changedTouches[0].clientY - this.startY < this.headHeight) {
        this.duration = this.animationDuration
        this.getStatus(0)
        return
      }
      // 松手回到默认位置
      this.duration = this.animationDuration
      this.getStatus(this.headHeight, true)
      this.$emit('refresh')
      this.$emit('input', true)
    },
    /**
     * 获取页面当前状态
     */
    getStatus(height, isLoading) {
      this.height = height
      const status = isLoading ? 'loading' : height === 0 ? 'normal' : height < this.headHeight ? 'pulling' : 'loosing'
      if (this.status !== status) {
        this.status = status
      }
    },
    /**
     * 获取页面滚动条高度
     */
    getScrollTop() {
      let scrollTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop
      return scrollTop
    }
  },
  watch: {
    value(nv) {
      this.duration = this.animationDuration
      this.getStatus(nv ? this.headHeight : 0, nv)
    }
  }
}
</script>
<style lang="scss" scoped>
.kn-pullrefresh {
  overflow: hidden;
  user-select: none; // 用户不能选中复制文本

  .track {
    position: relative;
  }

  .head {
    position: absolute;
    top: -50px;
    left: 0;
    width: 100%;
    height: 50px;
    overflow: hidden;
    font-size: 14px;
    line-height: 50px;
    color: gray;
    text-align: center;

    .hint {
      width: 100%;
      height: 50px;
      line-height: 50px;
      text-align: center;
    }
  }
}
</style>
