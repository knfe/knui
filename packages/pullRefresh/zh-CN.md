<!-- 简体中文 -->

# kn-pullrefresh
