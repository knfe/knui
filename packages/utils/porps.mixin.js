export const mixins = {
  data: () => {
    return {
      countDownTxt: '重新获取', // 倒计时文案
      countDownTime: 60,
      index: 'index' + new Date().getTime() // 编号
    }
  },
  props: {
    visible: {
      type: Boolean,
      default: () => {
        return false
      }
    }, // message是否可见(用于过渡)
    title: {
      type: String,
      default: () => {
        return ''
      }
    }, // 标题
    message: {
      type: String,
      default: () => {
        return ''
      }
    }, // message内容
    btn: {
      type: Array,
      default: () => {
        return []
      }
    }, // message按钮
    leftLoading: {
      type: Boolean,
      default: () => {
        return false
      }
    }, // 是否加载loading
    rightLoading: {
      type: Boolean,
      default: () => {
        return false
      }
    }, // 是否加载loading
    shadowClose: {
      type: Boolean,
      default: () => {
        return false
      }
    }, // 背景点击是否可以关闭loading
    className: {
      type: String,
      default: () => {
        return ''
      }
    }, // 自定义类名
    lockScroll: {
      type: Boolean,
      default: () => {
        return true
      }
    }, // 是否在 Message 出现时将 body 滚动锁定
    hasInput: {
      type: Boolean,
      default: () => {
        return true
      }
    },
    inputType: {
      type: String,
      default: () => {
        return 'text'
      }
    },
    inputMaxlength: {
      type: String,
      default: () => {
        return ''
      }
    },
    hasCountDown: {
      type: Boolean,
      default: () => {
        return true
      }
    },
    inputValue: {
      type: String,
      default: () => {
        return ''
      }
    }
  },
  methods: {
    close() {
      this.$emit('close')
    },
    leftClick() {
      if (typeof this.leftBtnClick !== 'function') {
        this.close()
        return
      }
      if (this.leftBtnCloseLoading) {
        this.leftLoading = true
      }
      this.leftBtnClick(this.inputValue)
    },
    rightClick() {
      if (typeof this.rightBtnClick !== 'function') {
        this.close()
        return
      }
      if (this.rightBtnCloseLoading) {
        this.rightLoading = true
      }
      this.rightBtnClick(this.inputValue)
    },
    closeShadow() {
      if (!this.shadowClose) return
      this.close()
    },
    countDown() {
      if (typeof this.countDownClick === 'function') {
        this.countDownClick()
      }
      let timer0
      if (timer0) clearInterval(timer0)
      let count = this.countDownTime

      this.countDownTxt = `已发送(${count}s)`
      timer0 = setInterval(() => {
        count--
        // console.log(count)
        this.countDownTxt = `已发送(${count}s)`
        if (count <= 0) {
          clearInterval(timer0)
          this.countDownTxt = `重新获取`
        }
      }, 1000)
    }
  }
}
