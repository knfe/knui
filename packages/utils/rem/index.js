export default function(baseWidth, baseRate) {
  if (baseWidth === void 0) {
    baseWidth = 375
  }
  if (baseRate === void 0) {
    baseRate = 100
  }
  function setHtmlFontSize() {
    var screenWidth = window.innerWidth
    // console.log(screenWidth, baseWidth, baseRate)
    var fontSize = (screenWidth / baseWidth) * baseRate
    // console.log(fontSize)
    document.documentElement.style.fontSize = fontSize + 'px'
  }
  setHtmlFontSize()
  window.onload = setHtmlFontSize
  window.onresize = setHtmlFontSize
}
