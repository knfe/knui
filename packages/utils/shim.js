/**
 * @function 兼容性处理函数
 * @author KeNo
 */

/**
 * 得到鼠标位置的方法
 * @param event 事件的名称
 * @returns {x: 距离左边的距离, y: 距离上方的距离}
 */
export function getPage(event) {
  var pageX = event.pageX || getScroll().left + event.clientX
  var pageY = event.pageY || getScroll().top + event.clientY
  return { x: pageX, y: pageY }
}

/**
 * 得到卷曲部位大小的方法
 * @returns {top: 上方卷曲的距离, left: 左方卷曲的距离}
 */
export function getScroll() {
  var scrollTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop
  var scrollLeft = window.pageXOffset || document.documentElement.scrollLeft || document.body.scrollLeft
  return { top: scrollTop, left: scrollLeft }
}

/**
 * 添加事件的方法
 * @param ele 事件源
 * @param type  事件名称 "需要引号"
 * @param fn  事件发生函数
 */
export function myAddEvent(ele, type, fn) {
  if (ele.addEventListener) {
    ele.addEventListener(type, fn)
  } else if (ele.attachEvent) {
    ele.attachEvent('on' + type, fn)
  } else {
    ele['on' + type] = fn
  }
}

/**
 * 删除事件的方法
 * @param ele 事件源
 * @param type  事件名称 "需要引号"
 * @param fn  事件发生函数
 */
export function myRemoveEvent(ele, type, fn) {
  if (ele.removeEventListener) {
    ele.removeEventListener(type, fn)
  } else if (ele.detachEvent) {
    ele.detachEvent('on' + type, fn)
  } else {
    ele['on' + type] = null
  }
}

/**
 * 获取非行内样式
 * @param ele   被获取的元素
 * @param attrName  需要获取的属性名
 * @returns     返回该属性的样式
 */
export function getStyle(ele, attrName) {
  if (window.getComputedStyle) {
    // 如果使用getComputedStyle能够拿到样式
    return window.getComputedStyle(ele, null)[attrName]
  }
  return ele.currentStyle[attrName]
}

/**
 * 获取可视区域的宽和高
 * @returns width和height
 */
export function getClient() {
  var width = window.innerWidth || document.documentElement.clientWidth || document.body.clientWidth || 0
  var height = window.innerHeight || document.documentElement.clientHeight || document.body.clientHeight || 0

  return { clientWidth: width, clientHeight: height }
}

export default {
  getPage,
  getScroll,
  myAddEvent,
  myRemoveEvent,
  getStyle,
  getClient
}
