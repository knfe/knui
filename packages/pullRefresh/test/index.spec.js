import { shallowMount } from '@vue/test-utils'
import pullRefresh from '..'
import { trigger, triggerDrag } from '../../../tests/utils'

test('change head content when pulling down', () => {
  const wrapper = shallowMount(pullRefresh, {
    propsData: {
      value: false
    }
  })

  wrapper.vm.$on('input', value => {
    wrapper.vm.value = value
  })

  const track = wrapper.find('.track')

  // pulling
  trigger(track, 'touchstart', 0, 0)
  trigger(track, 'touchmove', 0, 10)
  expect(wrapper).toMatchSnapshot()

  // loosing
  trigger(track, 'touchmove', 0, 100)
  expect(wrapper).toMatchSnapshot()

  // loading
  trigger(track, 'touchend', 0, 100)
  expect(wrapper).toMatchSnapshot()

  // still loading
  triggerDrag(track, 0, 100)
  expect(wrapper).toMatchSnapshot()

  expect(wrapper.emitted('input')).toBeTruthy()
  expect(wrapper.emitted('refresh')).toBeTruthy()

  // end loading
  wrapper.vm.value = false
  expect(wrapper).toMatchSnapshot()
})

test('pull a short distance', () => {
  const wrapper = shallowMount(pullRefresh, {
    propsData: {
      value: false
    }
  })

  const track = wrapper.find('.track')
  triggerDrag(track, 0, 10)
  expect(wrapper.emitted('input')).toBeFalsy()
})

test('not in page top', () => {
  const wrapper = shallowMount(pullRefresh, {
    propsData: {
      value: false
    }
  })

  window.scrollTop = 100

  const track = wrapper.find('.track')
  // ignore touch event when not at page top
  triggerDrag(track, 0, 100)
  window.scrollTop = 0
  trigger(track, 'touchmove', 0, 100)

  expect(wrapper).toMatchSnapshot()
})
