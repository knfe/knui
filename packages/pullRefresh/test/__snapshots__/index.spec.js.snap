// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change head content when pulling down 1`] = `
<div class="kn-pullrefresh">
  <div class="track" style="transform: translate3d(0,20px, 0);">
    <div class="head" style="height: 50px; line-height: 50px; top: -50px;">
      <!---->
      <div class="hint" style="height: 50px; line-height: 50px;">下拉即可刷新</div>
      <!---->
      <!---->
    </div>
  </div>
</div>
`;

exports[`change head content when pulling down 2`] = `
<div class="kn-pullrefresh">
  <div class="track" style="transform: translate3d(0,63.245553203367585px, 0);">
    <div class="head" style="height: 50px; line-height: 50px; top: -50px;">
      <!---->
      <!---->
      <div class="hint" style="height: 50px; line-height: 50px;">释放即可刷新</div>
      <!---->
    </div>
  </div>
</div>
`;

exports[`change head content when pulling down 3`] = `
<div class="kn-pullrefresh">
  <div class="track" style="transform: translate3d(0,50px, 0);">
    <div class="head" style="height: 50px; line-height: 50px; top: -50px;">
      <!---->
      <!---->
      <!---->
      <div class="hint" style="height: 50px; line-height: 50px;">正在加载...</div>
    </div>
  </div>
</div>
`;

exports[`change head content when pulling down 4`] = `
<div class="kn-pullrefresh">
  <div class="track" style="transform: translate3d(0,50px, 0);">
    <div class="head" style="height: 50px; line-height: 50px; top: -50px;">
      <!---->
      <!---->
      <!---->
      <div class="hint" style="height: 50px; line-height: 50px;">正在加载...</div>
    </div>
  </div>
</div>
`;

exports[`change head content when pulling down 5`] = `
<div class="kn-pullrefresh">
  <div class="track" style="transform: translate3d(0,0px, 0);">
    <div class="head" style="height: 50px; line-height: 50px; top: -50px;">
      <!---->
      <!---->
      <!---->
    </div>
  </div>
</div>
`;

exports[`not in page top 1`] = `
<div class="kn-pullrefresh">
  <div class="track" style="transform: translate3d(0,50px, 0);">
    <div class="head" style="height: 50px; line-height: 50px; top: -50px;">
      <!---->
      <!---->
      <!---->
      <div class="hint" style="height: 50px; line-height: 50px;">正在加载...</div>
    </div>
  </div>
</div>
`;
